atomix/cluster/src/test/java/io/atomix/raft/utils/VoteQuorumTest.java

```diff
@@ -0,0 +1,249 @@
+              callback, Set.of(MemberId.from("1"), MemberId.from("2"), MemberId.from("3")));
+
+      // when
+      quorum.succeed(new MemberId("5"));
+      quorum.succeed(new MemberId("6"));
+
+      // then
+      verifyNoInteractions(callback);
+    }
+
+    @Test
+    void shouldOnlySucceedOnce() {
+      // given
+      final Consumer<Boolean> callback = mock();
+
+      final var quorum =
+          new SimpleVoteQuorum(
+              callback, Set.of(MemberId.from("1"), MemberId.from("2"), MemberId.from("3")));
+
+      // when
+      quorum.succeed(new MemberId("1"));
+      quorum.succeed(new MemberId("2"));
+      quorum.succeed(new MemberId("3"));
+
+      // then
+      verify(callback, only()).accept(true);
+    }
+
+    @Test
+    void shouldFail() {
+      // given
+      final Consumer<Boolean> callback = mock();
+
+      final var quorum =
+          new SimpleVoteQuorum(
+              callback, Set.of(MemberId.from("1"), MemberId.from("2"), MemberId.from("3")));
+
+      // when
+      quorum.fail(new MemberId("1"));
+      quorum.fail(new MemberId("2"));
+
+      // then
+      verify(callback, only()).accept(false);
+    }
+
+    @Test
+    void shouldNotRecoverFromFail() {
+      // given
+      final Consumer<Boolean> callback = mock();
+
+      final var quorum =
+          new SimpleVoteQuorum(
+              callback, Set.of(MemberId.from("1"), MemberId.from("2"), MemberId.from("3")));
+
+      // when
+      quorum.fail(new MemberId("1"));
+      quorum.fail(new MemberId("2"));
+      quorum.succeed(new MemberId("1"));
+      quorum.succeed(new MemberId("2"));
+
+      // then
+      verify(callback, only()).accept(false);
+    }
+  }
+
+  @Nested
+  final class JointConsensus {
+    @Test
+    void shouldSucceedWithOverlappingMembers() {
+      // given
+      final Consumer<Boolean> callback = mock();
+
+      final var quorum =
+          new JointConsensusVoteQuorum(
+              callback,
+              Set.of(MemberId.from("1"), MemberId.from("2"), MemberId.from("4")),
+              Set.of(MemberId.from("1"), MemberId.from("2"), MemberId.from("3")));
+
+      // when
+      quorum.succeed(new MemberId("1"));
+      quorum.succeed(new MemberId("2"));
+
+      // then
+      verify(callback, only()).accept(true);
+    }
+
+    @Test
+    void shouldSucceedWithDisjointMembers() {
+      // given
+      final Consumer<Boolean> callback = mock();
+
+      final var quorum =
+          new JointConsensusVoteQuorum(
+              callback,
+              Set.of(MemberId.from("1"), MemberId.from("2"), MemberId.from("3")),
+              Set.of(MemberId.from("4"), MemberId.from("5"), MemberId.from("6")));
+
+      // when
+      quorum.succeed(new MemberId("1"));
+      quorum.succeed(new MemberId("2"));
+      quorum.succeed(new MemberId("4"));
+      quorum.succeed(new MemberId("5"));
+
+      // then
+      verify(callback, only()).accept(true);
+    }
+
+    @Test
+    void shouldWaitWithPartialQuorum() {
```

Comment:
:wrench: Also add tests where oldMembers have quorum and newMembers does not, and vise versa.